### DISCUSSION TITLE
Official documents are really hard to understand

### DISCUSSION THREAD

**Discussion by Jzow** [diesel-rs/diesel#3162] [Q&A]
At present, part of the code of my project has been implemented through MySQL primitives. I recently saw ORM framework diesel. I tried to see some examples on the official website page, which I found difficult to understand. Then I looked at the MySQL examples in the example module. In addition to the link database, I can understand, such as schma I don't quite understand

**Comment by weiznich** [comment]
I've moved this to the discussion forum, as this does not follow the bug report template. (That's mandatory for our issue tracker.)

You may want to add more details about which parts of which guides are hard to understand for you. Maybe even give an example, or link the relevant parts. Otherwise this is not really actionable for anyone and you will likely not receive much help here.

**Comment by Jzow** [comment]
[getting-started](https://diesel.rs/guides/getting-started)

```rust
extern crate diesel_demo;
extern crate diesel;

use self::diesel_demo::*;
use self::models::*;
use self::diesel::prelude::*;

fn main() {
    use diesel_demo::schema::posts::dsl::*;

    let connection = establish_connection();
    let results = posts.filter(published.eq(true))
        .limit(5)
        .load::<Post>(&connection)
        .expect("Error loading posts");

    println!("Displaying {} posts", results.len());
    for post in results {
        println!("{}", post.title);
        println!("----------\n");
        println!("{}", post.body);
    }
}

```

Where did published come from? Where does Post come from? This post made me mistake it for actix_web Post

**Reply by weiznich** [reply]
Thanks for providing this context.

The code snipped you've posted already contains the relevant information via the corresponding use statements on top of the snippet as rust generally does not bring items in scope magically without an explicit import. Checkout the [general rust documentation](https://doc.rust-lang.org/stable/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html) on the modules and use for more information about how imports work.
The corresponding modules can be viewed by clicking on the "View on Github" link. Additionally the code of the corresponding other modules is given in the code blocks literally above the linked code block.

That all written: Do you have suggestions how this could be made more clear?

**Reply by amitu** [reply]
I have a suggestion:

I generally avoid `use` statements. When reviewing code in Github PR etc, I do not want to do mental gymnastics to understand a line of code, just jump to the top of file by clicking on the "view file" in Github is major hinderance. Code is written once, and read thousands of times.

I feel for code snippets like this, in docs, its particularly bad. In case of docs, docs is written once and read millions of time maybe. I feel code snippets should be easy to copy paste, right now in this snippet it requires too much work to know where `establish_connection()` comes from for example. 

This is an code snippet from a project of mine:

```rs
pub fn translation_follower(cin: &Cache, user_id: i32, followee: i32, follower: i32) -> Result<()> {
    use crate::schema::fifthtry_cf;

    diesel::insert_into(fifthtry_cf::table)
        .values((
            fifthtry_cf::relation.eq("translation"),
            fifthtry_cf::order.eq(1),
            fifthtry_cf::auto_merge.eq(true),
            fifthtry_cf::created_on.eq(cin.in_.now),
            fifthtry_cf::updated_on.eq(cin.in_.now),
            fifthtry_cf::follower.eq(follower),
            fifthtry_cf::followee.eq(followee),
            fifthtry_cf::created_by.eq(user_id),
        ))
        .execute(cin.in_.conn)
        .map(|_| ())
        .map_err(Into::into)
}
```

Notice I do not use `use diesel_demo::schema::posts::dsl::*;`. I do have `use diesel::prelude::*`, to bring in some traits, and I would have preferred if there was something like `use diesel::only_the_minimal_traits_to_make_diesel_work::*`.

**Reply by weiznich** [reply]
> Notice I do not use `use diesel_demo::schema::posts::dsl::*;`

That's a good idea. I'm happy to receive PR's for this.

>  I do have use diesel::prelude::*, to bring in some traits, and I would have preferred if there was something like use diesel::only_the_minimal_traits_to_make_diesel_work::*.

That's unfortunately not as simple as it sounds. Depending on your query you might need to use one or the other trait from diesel. That's especially true around certain `*ExpressionMethods` traits. `diesel::prelude::*` is already more or less the minimal amount of imports which are required to support all expressable queries. At least without having wired edge cases like, if you would like to use this method you need to import another trait.

**Comment by Jzow** [comment]
```rust
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize, Queryable)]
pub struct InstitutionObject {
    #[serde(default)]
    pub id: Option<String>,
    #[serde(default)]
    pub institution_name: Option<String>,
    #[serde(default)]
    pub institution_address: Option<String>,
    #[serde(default)]
    pub institution_phone: Option<String>,
    #[serde(default)]
    pub institution_region: Option<String>,
    #[serde(default)]
    pub create_time: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub update_time: Option<String>,
}

   type NucleicInstitution = nucleic_institution::InstitutionObject;

    fn query_nucleic_institution_byregion(region: &str) -> bool {

        use schema::nucleic_test_institution::dsl::*;
        
        let coon = diesel_dbconfig::mysql_connection();

        let result = nucleic_test_institution
            .filter(institution_region.eq(region))
            .limit(10)
            .load::<NucleicInstitution>(&coon)
            .expect("Error loading posts");

        false
    }
```


load error:
```
the trait bound `std::option::Option<std::string::String>: diesel::Queryable<diesel::sql_types::Text, _>` is not satisfied
the following implementations were found:
  <std::option::Option<T> as diesel::Queryable<diesel::sql_types::Nullable<ST>, DB>>
required because of the requirements on the impl of `diesel::Queryable<(diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text), _>` for `(std::option::Option<std::string::String>, std::option::Option<std::string::String>, std::option::Option<std::string::String>, std::option::Option<std::string::String>, std::option::Option<std::string::String>, std::option::Option<std::string::String>, std::option::Option<std::string::String>)`
1 redundant requirement hidden
required because of the requirements on the impl of `diesel::Queryable<(diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text, diesel::sql_types::Text), _>` for `InstitutionObject`
required because of the requirements on the impl of `LoadQuery<_, InstitutionObject>` for `diesel::query_builder::SelectStatement<table, query_builder::select_clause::DefaultSelectClause, query_builder::distinct_clause::NoDistinctClause, query_builder::where_clause::WhereClause<diesel::expression::operators::Eq<columns::institution_region, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>>, query_builder::order_clause::NoOrderClause, query_builder::limit_clause::LimitClause<diesel::expression::bound::Bound<BigInt, i64>>>`rustc[E0277](https://doc.rust-lang.org/error-index.html#E0277)
mod.rs(1238, 15): required by a bound in `load`
```

**Reply by weiznich** [reply]
This error message indicates a type mismatch between one of your struct fields and what your query returns. It seems like all fields returned by your query are not nullable, but you use `Option` in your struct everywhere.

**Reply by Jzow** [reply]
Yes, I can also set it to be empty when inserting. This needs to modify the **table!**

**Reply by weiznich** [reply]
Again: That's the answer. It will not change just because you don't like it or you imaging things work different. Diesel is quite strict about mapping types, similar to rust. You cannot assign a `i32` to a `Option<i32>` without explicitly wrapping it into `Some` in rust either. Therefore diesel does not allow you to mix nullable and non nullable types as well. They are just different types. 

Generally speaking: Please consider to take more time to formulate your questions in such a way that they are understandable by other people. This should include more than just dumping incomplete code parts and a part of an error messages. I will likely stop responding to such posts.
